### PR TITLE
feat(signals): prevent overriding of SignalStore members

### LINKED ISSUE #4144: [signals] Prevent overriding state fields with a computed signal of the same name
### Which @ngrx/* package(s) are the source of the bug?

signals

### Minimal reproduction of the bug/regression with instructions

I'm not sure if this is intentional or not: it's possible to define a computed field with the same name as a field already defined in the initial state.

**Reproduction:**

https://stackblitz.com/edit/stackblitz-starters-atpbi6?file=src%2Fmain.ts

For posterity, here's the store definition used in the StackBlitz above:

```ts
const Store = signalStore(
  { providedIn: 'root' },
  withState({ value: 1 }),
  withComputed((store) => ({
    value: computed(() => store.value() * 2),
  }))
);
```

This may lead to unintentional bugs.

### Expected behavior

You should get an error if you try to define a computed field with same name as an existing field in the initial state.

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

Angular: v17
NgRx: v17.0.0

### Other information

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #4144

It's allowed to override SignalStore members:

```ts
const DemoStore = signalStore({
  withState({ foo: 'bar' }),
  withMethods(({ foo }) => ({ // `foo` type: Signal<string>
    foo() {
      console.log(foo());
    }
  }),
  withHooks({
    onInit({ foo }) { // `foo` type: () => void
      foo();
    },
  })
});
```

## What is the new behavior?

- The warning will be displayed in development mode if a user tries to override SignalStore members.
- `StateSignal` is renamed to `StateSource` for clarity - avoiding confusion between `StateSignal` and `SignalState`.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information

After this change, signal stores that rely on overriding have to be changed. Also, if `StateSignal` is used, it should be replaced with `StateSource`:

```diff
-import { patchState, StateSignal } from '@ngrx/signals';
+import { patchState, StateSource } from '@ngrx/signals';

-resetCount(state: StateSignal<{ count: number }>): void {
+resetCount(state: StateSource<{ count: number }>): void {
  patchState(state, { count: 0 });
}
```

This is not considered as a breaking change because the `@ngrx/signals` package is in developer preview.

### CODE REVIEW

File: modules/signals/src/signal-store-assertions.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { InnerSignalStore } from './signal-store-models';
+
+declare const ngDevMode: unknown;
+
+export function assertUniqueStoreMembers(
+  store: InnerSignalStore,
+  newMemberKeys: string[]
+): void {
+  if (!ngDevMode) {
```
Comment (brandonroberts) [inline modules/signals/src/signal-store-assertions.ts:9]: Can we wrap this whole check inside the `ngDevMode` block so it gets tree-shaken during the prod build?

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: modules/signals/src/signal-store-assertions.ts (hunk 1 above)
Comment (markostanimirovic) [reply modules/signals/src/signal-store-assertions.ts:9]: Tree-shaking works with existing implementation as well. Tested on `standalone-app`.

**Review by brandonroberts — APPROVED** [review]

File: modules/signals/src/index.ts [hunk 2]
```diff
@@ -1,11 +1,14 @@
...
+export { signalState, SignalState } from './signal-state';
 export { signalStore } from './signal-store';
 export { signalStoreFeature, type } from './signal-store-feature';
 export { SignalStoreFeature } from './signal-store-models';
-export { StateSignal } from './state-signal';
+export {
+  getState,
+  PartialStateUpdater,
+  patchState,
+  StateSource,
```
Comment (timdeschryver) [inline modules/signals/src/index.ts:10]: Unrelated to the PR, but do we want to have migration in place to update StateSignal to StateSource ?

**Review by timdeschryver — APPROVED** [review]

File: modules/signals/src/index.ts (hunk 2 above)
Comment (markostanimirovic) [reply modules/signals/src/index.ts:10]: yes, that would be great 👍

**Pull request merged**